### ISSUE TITLE
C++ SysV ABI incompatibility wrt. passing non-PODs by value

### ISSUE THREAD

**Issue by kinke** [ldc-developers/ldc#2702]
While checking assembly and clang-IR wrt. [DIP 1014](https://forum.dlang.org/thread/zfyfhqczkrfdpfkcabna@forum.dlang.org), I stumbled on what looks like a huge ABI incompatibility, at least for the SysV ABI, and probably the 32-bit x86 ABIs too (Win64 should be safe though).

As soon as a C++ struct/class has a copy or move ctor, pass-by-value semantics seem to match `ExplicitByvalRewrite` (pass a pointer to a caller-allocated copy). LDC and DMD instead pass the aggregate on the stack (i.e., like a big POD struct). More information to follow.

**Comment by kinke** [comment]
To play around with C++: https://godbolt.org/g/XV8Pem (try commenting out the copy ctor and see what happens)

**Comment by kinke** [comment]
GDC, at least v5.2 as latest version on godbolt, is affected too. See this [D snippet](https://d.godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(j:1,lang:___d,source:'extern(C%2B%2B)+struct+S%0A%7B%0A++++size_t+member%3B%0A++++this(size_t+m)+%7B+member+%3D+m%3B+%7D%0A++++//this(this)%3B%0A++++//~this()%3B%0A++++//this(const+ref+S+rhs)%3B%0A%7D%3B%0A%0Asize_t+foo(S+param)+%7B+return+param.member+%2B+123%3B+%7D%0A%0Avoid+bar()%0A%7B%0A++++foo(S(1))%3B%0A%0A++++S+lvalue+%3D+S(2)%3B%0A++++foo(lvalue)%3B%0A%0A++++//import+std.algorithm.mutation+:+move%3B%0A++++//foo(move(lvalue))%3B%0A%7D%0A'),l:'5',n:'0',o:'D+source+%231',t:'0')),k:32.87843592721642,l:'4',m:100,n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:gdc52,filters:(b:'0',binary:'1',commentOnly:'0',demangle:'0',directives:'0',execute:'1',intel:'0',trim:'0'),lang:___d,libs:!(),options:'',source:1,wantOptInfo:'1'),l:'5',n:'0',o:'gdc+5.2.0+(Editor+%231,+Compiler+%231)+D',t:'0')),k:33.78823073945026,l:'4',m:100,n:'0',o:'',s:0,t:'0'),(g:!((h:output,i:(compiler:1,editor:1,wrap:'1'),l:'5',n:'0',o:'%231+with+gdc+5.2.0',t:'0')),k:33.33333333333333,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4).

D additionally doesn't call the C++ copy ctor when passing an lvalue by value, I thought that'd be implemented already. That can be worked around with a D postblit ctor though.

Then there's the different destruction rules; D moves the arg (or arg copy for lvalues) to the callee's memory params stack and lets the callee destruct it, C++ passes a ref to the rvalue (or lvalue copy) to the callee and the calleR destructs it after the call. In frontend terms, a C++ non-POD by-value param is more like a special ref, accepting rvalue args and requiring a copy for lvalue args.

Anyway, this explains why clang has no problems calling a move ctor, while LDC would probably have a hard time with the current ABI (final address of callee param only exposed to callee via implicit `byval` param; original address of moved-from arg only known by caller).
So DIP 1014 
In short, there's still a fair amount of work required to support passing non-PODs by value across language boundaries.

**Comment by kinke** [comment]
Non-PODs use the `IndirectByvalRewrite` now, and the C++ destruction rules are applied now too since v2.096 (https://issues.dlang.org/show_bug.cgi?id=20235).
